### PR TITLE
daemon: lower allocations

### PR DESCRIPTION
**Pull request by unclejack**
This PR helps improve things with #12899. I'm not going to say this fully fixes it for now as there are other changes to be made.

### CODE REVIEW

**Comment by LK4D4** [comment]
@unclejack Could you describe how it helps save memory? Also #12899 definitely has origins in never-removed execConfig structs.

**Comment by unclejack** [comment]
@LK4D4 execConfig structs are moste likely a small part of this. This saves memory by avoiding the allocation of io.Copy buffers. This saves about 96 KB per exec.

**Comment by LK4D4** [comment]
Makes sense to write own `Copy` function for this.

**Comment by unclejack** [comment]
@LK4D4 Fair enough, I'll add it to pools.

File: daemon/exec.go [hunk 1]
```diff
@@ -183,7 +184,9 @@ func (d *Daemon) ContainerExecStart(execName string, stdin io.ReadCloser, stdout
 		go func() {
 			defer w.Close()
 			defer logrus.Debugf("Closing buffered stdin pipe")
-			io.Copy(w, stdin)
+			buf := pools.BufioReader32KPool.Get(stdin)
+			defer pools.BufioReader32KPool.Put(buf)
+			io.Copy(w, buf)
```
Comment (stevvooe) [inline daemon/exec.go]: I know this is unrelated to the PR, but could we add a check for this error here.

**Comment by stevvooe** [comment]
@LK4D4 Sorry to disagree with you here, but we should avoid having a specialized version of `io.Copy`. It is a subtle function to get right. `io.Copy` can defer buffer allocation with `ReaderFrom` and `WriterTo`. Perhaps, a helper method like `BufferedReaderFromSize(rd, size)` would be appropriate.

**Comment by LK4D4** [comment]
@stevvooe but our function will be just wrapper around. Helper method not helping here, because we have hardcoded pools in pkg/pools.

**Comment by unclejack** [comment]
@LK4D4 It's done. I've added it as a package level exported function for convenience.
@stevvooe As @LK4D4, it was only going to be a wrapper, not a full implementation.

**Comment by LK4D4** [comment]
@unclejack Looks like some tests are broken.

**Comment by unclejack** [comment]
@LK4D4 I've reproduced the failure and I'm working on fixing it right away.

**Comment by unclejack** [comment]
@LK4D4 The tests seem to be OK now. There's more work to be done for this, but this is good enough for one PR.

**Comment by calavera** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**